docs/errors.md

```diff
@@ -0,0 +1,249 @@
+    - [`MongoRuntimeError`](#MongoRuntimeError)
+  - [`MongoNetworkError`](#MongoNetworkError)
+  - [`MongoServerError`](#MongoServerError)
+  - [`MongoSystemError`](#MongoSystemError)
+- [Test Plan](#Test-Plan)
+  - [`MongoLogicError`](#MongoLogicError-1)
+    - [`MongoInvalidArgumentError`](#MongoInvalidArgumentError-1)
+    - [`MongoClientInstantiationError`](#MongoClientInstantiationError-1)
+    - [`MongoMissingCredentialsError`](#MongoMissingCredentialsError-1)
+  - [`MongoRuntimeError`](#MongoRuntimeError-1)
+    - [`MongoClientNotConnectedError`](#MongoClientNotConnectedError-1)
+    - [`MongoServerClosedError`](#MongoServerClosedError-1)
+    - [`MongoStreamClosedError`](#MongoStreamClosedError-1)
+    - [`MongoTopologyClosedError`](#MongoTopologyClosedError-1)
+    - [`MongoCursorExhaustedError`](#MongoCursorExhaustedError-1)
+    - [`MongoIOError`](#MongoIOError-1)
+  - [`MongoNetworkError`](#MongoNetworkError-1)
+    - [`MongoNetworkTimeoutError`](#MongoNetworkTimeoutError-1)
+
+# Errors
+
+All errors are derived from the `MongoError` class which should **never** be instantiated.
+There are four main error classes which stem from `MongoError`: `MongoDriverError`, `MongoNetworkError`, `MongoServerError`, and `MongoSystemError`.
+
+## `MongoError`
+
+The base class from which all errors in the Node driver subclass.
+`MongoError` should **never** be be directly instantiated.
+![(MongoError hierarchy tree)](charts/imgs/MongoError.svg)
+
+Children of `MongoError` include:
+
+- [`MongoDriverError`](#MongoDriverError)
+- [`MongoNetworkError`](#MongoNetworkError)
+- [`MongoServerError`](#MongoServerError)
+- [`MongoSystemError`](#MongoSystemError)
+
+## `MongoDriverError`
+
+This class represents errors which originate in the driver itself or in the user's use of the driver.
+This class should **never** be directly instantiated.
+Its children are the main classes of errors that most users will interact with: [**`MongoLogicError`**](#MongoLogicError) and [**`MongoRuntimeError`**](#MongoRuntimeError).
+
+### `MongoLogicError`
+
+This class represents errors which originate from misuse of the driver API and will generally be thrown before making contact with the server.
+This class should **never** be directly instantiated.
+![(MongoLogicError hierarchy tree)](charts/imgs/MongoLogicError.svg)
+Children of `MongoLogicError` include:
+
+- #### `MongoInvalidArgumentError`
+
+  - Thrown when the user supplies malformed, unexpected arguments or failed to provide a required argument or field.
+
+- #### `MongoCompatibilityError`
+
+  - Thrown when a feature that is not enabled or allowed for the current configuration is used.
+
+- #### `MongoClientInstantiationError`
+
+  - Thrown when a user supplies a poorly-formatted connection string to the MongoClient constructor.
+
+- #### `MongoMissingCredentialsError`
+
+  - Thrown when a user fails to provide authentication credentials before attempting to connect to the mongo server.
```

Comment:
What's the use case where this error would arise? you can run mongodb without authentication, but there isn't a way to tell that's the case until you attempt to connect.

Reply:
Looking through some of the files, we thought this error would be used when Mongo was paired with different authentication methods like x509 or SCRAM. Almost all of the auth mechanisms under `src/cmap/auth` seem to have a generic must provide credentials error, so this class would unify that.

Reply:
Oh I see yes that seems like a good classification, although out of scope from this work I wonder if it ever gets used, I believe our validation is strict enough such that you would be warned at client construction time that something is not setup right. That being said lets go ahead and keep this